### PR TITLE
Initial Highlight editing in SummaryPopup

### PR DESCRIPTION
**Pull request by PiotrKozlowski**
for: https://app.zenhub.com/workspaces/openstax-unified-5b71aabe3815ff014b102258/issues/openstax/unified/588

Initial editing menu for highlights in summary popup

TODO:
- [x] Check on mobile
- [x] Close context menu on click outside
~~- [ ] Add tests (After merging filters pr and updating this branch with master)~~ I'll do this in separate pr for easier review

### CODE REVIEW

File: src/app/content/highlights/components/Highlights.tsx
Comment (TomWoodward) [inline src/app/content/highlights/components/Highlights.tsx]: lets move all this logic down into a per-highlight component. the only reason i can think of to have the logic in the list is to do some special handling for editing or blocking the editing of multiple highlights at once, but suhani has said we don't need to do that

File: src/app/content/highlights/components/Highlights.tsx
Comment (TomWoodward) [inline src/app/content/highlights/components/Highlights.tsx]: if this isn't in the spec (i don't think it is?) please don't do any special enter key handling

File: src/app/content/highlights/components/Highlights.tsx
Comment (TomWoodward) [inline src/app/content/highlights/components/Highlights.tsx]: handling the esc key is nice but is actually a much larger problem we need to address with using it to close dropdowns, dialogs, i would not add this for now and wait until we address the problem as a whole

File: src/app/content/highlights/components/Highlights.tsx
Comment (TomWoodward) [inline src/app/content/highlights/components/Highlights.tsx]: i think you split these components into new files in a different PR, but you can go ahead and do it on this branch, it would be easier to read with them in different files

**Review by TomWoodward — CHANGES_REQUESTED** [review]

File: src/app/content/highlights/components/Highlights.tsx
Comment (PiotrKozlowski) [reply src/app/content/highlights/components/Highlights.tsx]: ok, sorry, I'm just used to working without design specs when review is ending at: please implement also this, this and this :D

File: src/app/content/highlights/components/ShowMyHighlightsStyles.tsx
Comment (TomWoodward) [inline src/app/content/highlights/components/ShowMyHighlightsStyles.tsx]: i usually specify all of top/bottom/left/right to 0 instead of with/height 100 to take up the whole container, do you know if there is a difference in behavior? i think this is fine i'm just curious.

File: src/app/content/highlights/components/ShowMyHighlightsStyles.tsx
Comment (TomWoodward) [inline src/app/content/highlights/components/ShowMyHighlightsStyles.tsx]: since `HighlightDeleteWrapper` has its own file, i would put the styles for it in there

File: src/app/content/highlights/reducer.ts
Comment (TomWoodward) [inline src/app/content/highlights/reducer.ts]: i'm not sure this logic is working correctly. i think you should process the current page highlights and the summary highlights separately, and each one should be in a function that correctly noops and returns the input state if no action is required. 

`updateSummaryHighlightsDependOnFilters` it has a noop if the chapter doesn't match filters, but doesn't have one for changing a color that did not match filters to another color that also doesn't match filters.


edit: the `updateHighlights` function does this already, it returns the input state if no action is necessary, so there is no reason to separately check to see if action is necessary on its behalf.

File: src/app/content/highlights/utils/getHighlightByIdFromHighlights.ts
Comment (TomWoodward) [inline src/app/content/highlights/utils/getHighlightByIdFromHighlights.ts]: this function seems excessively complicated. before the index was being fetched so that it could then be used to write back in the new highlight object, in this case there is no reason to findIndex you could use the regular find. you're also copying the object here which is undocumented and unnecessary

File: src/app/content/highlights/utils/summaryHighlightsUtils.ts
Comment (TomWoodward) [inline src/app/content/highlights/utils/summaryHighlightsUtils.ts]: again, a regular find would work here

File: src/app/content/highlights/utils/summaryHighlightsUtils.ts
Comment (TomWoodward) [inline src/app/content/highlights/utils/summaryHighlightsUtils.ts]: just return the highlight, no need for all the breaks

**Review by TomWoodward — CHANGES_REQUESTED** [review]

File: src/app/content/highlights/components/HighlightData.tsx
Comment (PiotrKozlowski) [reply src/app/content/highlights/components/HighlightData.tsx]: If I didn't pass color when I was updating annotation then color was removed but it worked properly in this case. However now both cases works properly (I can update annotation or color alone). Something has changed?

File: src/app/content/highlights/components/HighlightDeleteWrapper.tsx
Comment (PiotrKozlowski) [reply src/app/content/highlights/components/HighlightDeleteWrapper.tsx]: I'm (pr with tests) but I can target those buttons differently. I've just saw in other components that this is how it is used

File: src/app/content/highlights/components/HighlightToggleEdit.tsx
Comment (PiotrKozlowski) [reply src/app/content/highlights/components/HighlightToggleEdit.tsx]: Lets wait for the answer
https://openstax.slack.com/archives/C8V1XET1R/p1579095008268700

File: src/app/content/highlights/components/ShowMyHighlightsStyles.tsx
Comment (PiotrKozlowski) [reply src/app/content/highlights/components/ShowMyHighlightsStyles.tsx]: Those are smaller in specs: https://projects.invisionapp.com/d/main?origin=v7#/console/18130098/376272522/preview?scrollOffset=468

File: src/app/content/highlights/components/ShowMyHighlightsStyles.tsx
Comment (PiotrKozlowski) [reply src/app/content/highlights/components/ShowMyHighlightsStyles.tsx]: I don't know, I'm just used to using this way

File: src/app/content/highlights/components/HighlightData.tsx
Comment (PiotrKozlowski) [reply src/app/content/highlights/components/HighlightData.tsx]: Should I move also `Highlights.tsx` and `ShowMyHighlights.tsx` since those should be there also, but they will make huge diff for you to check

File: src/app/content/highlights/components/ShowMyHighlightsStyles.tsx
Comment (PiotrKozlowski) [reply src/app/content/highlights/components/ShowMyHighlightsStyles.tsx]: oh, I see, there is props.size to set

File: src/app/content/highlights/components/ShowMyHighlightsStyles.tsx
Comment (PiotrKozlowski) [reply src/app/content/highlights/components/ShowMyHighlightsStyles.tsx]: I'm just used to this, but I've spitted them

File: src/app/content/highlights/reducer.ts
Comment (PiotrKozlowski) [reply src/app/content/highlights/reducer.ts]: I'll add test for this case in #424 pr when this pr become approved

File: src/app/content/highlights/components/HighlightData.tsx
Comment (TomWoodward) [reply src/app/content/highlights/components/HighlightData.tsx]: i have not moved them because i'm afraid of conflicts with people, they really should be moved, but i'm not in a hurry

File: src/app/content/highlights/components/HighlightDeleteWrapper.tsx
Comment (TomWoodward) [reply src/app/content/highlights/components/HighlightDeleteWrapper.tsx]: i like using the test ids, i just want to make sure they're actually used

File: src/app/content/highlights/components/ShowMyHighlights.tsx
Comment (TomWoodward) [inline src/app/content/highlights/components/ShowMyHighlights.tsx]: lets try to keep this short
```suggestion
        data-analytics-region='MH popup'
```

File: src/app/content/highlights/components/ShowMyHighlightsStyles.tsx
Comment (TomWoodward) [inline src/app/content/highlights/components/ShowMyHighlightsStyles.tsx]: can you set this to `hidden` or `visible` or whatever you want here

File: src/app/content/highlights/components/SummaryPopup/HighlightDeleteWrapper.tsx
Comment (TomWoodward) [inline src/app/content/highlights/components/SummaryPopup/HighlightDeleteWrapper.tsx]: i like these short button labels, but to keep them in context we probably want 
`data-analytics-region="MH delete"` on the `StyledHighlightDeleteWrapper`

File: src/app/content/highlights/components/SummaryPopup/HighlightListElement.tsx
Comment (TomWoodward) [inline src/app/content/highlights/components/SummaryPopup/HighlightListElement.tsx]: this color should be `HighlightUpdateColorEnum` or whatever it actually is like `HighlightColorEnum`

File: src/app/content/highlights/utils/updateHighlights.ts
Comment (TomWoodward) [inline src/app/content/highlights/utils/updateHighlights.ts]: is this util used?

**Review by TomWoodward — CHANGES_REQUESTED** [review]
looking really great in general, left some minor comments, also needs conflicts resolved

File: src/app/content/highlights/components/SummaryPopup/HighlightDeleteWrapper.tsx
Comment (TomWoodward) [reply src/app/content/highlights/components/SummaryPopup/HighlightDeleteWrapper.tsx]: i realized that this is the second region i'm asking you to add for the same section of code, really both regions are good to have, i'm sorry the repeated comments are probably annoying though

File: src/app/content/highlights/components/SummaryPopup/HighlightListElement.tsx
Comment (PiotrKozlowski) [reply src/app/content/highlights/components/SummaryPopup/HighlightListElement.tsx]: I've updated ContextMenu and ColorPicker components to use `HighlightColorEnum` so it is more clear conversion

**Comment by TomWoodward** [comment]
i think this is good, but it needs master merged in and conflicts resolved

**Comment by PiotrKozlowski** [comment]
I think I've resolved conflicts, but I can't test it since highlight api is not working

**Comment by TomWoodward** [comment]
some of the snaps need updating @PiotrKozlowski

**Comment by PiotrKozlowski** [comment]
@TomWoodward I've updated this branch and `add-tests-to-588`

**Comment by TomWoodward** [comment]
@PiotrKozlowski i think this branch's code is fine but we can't merge. is your intent to merge #424 into this branch first?

**Review by TomWoodward — APPROVED** [review]

**Comment by TomWoodward** [comment]
waiting for qa/product review before merging

**Review by TomWoodward — APPROVED** [review]

**Comment by TomWoodward** [comment]
@PiotrKozlowski please look at the fixes i added, i'm submitting to qa now

**Comment by PiotrKozlowski** [comment]
@TomWoodward I've updated snapshots and according to https://github.com/openstax/unified/issues/588#issuecomment-586110403 we shouldn't add anything more in this pr, right?

**Pull request merged**
